### PR TITLE
fix maximum call stack size reached issue

### LINKED ISSUE #14744: Unable to open 'SQLQuery_1 - sqldev...grated)': Maximum call stack size exceeded.
- Azure Data Studio Version:

Steps to Reproduce:

1. Run following query until complete

declare @cnt int = 1
while @cnt <= 130000
begin
print @cnt
set @cnt = @cnt + 1
end

2. Open a new Query tab window

3. Return to the first query window, that is when you will receive the following error:
    "Unable to open 'SQLQuery_1 - sqldev...grated)': Maximum call stack size exceeded."

4. Also it will erase the SQLQuery_1 text
Azure Data Studio

### PR DESCRIPTION
**Pull request by alanrenmsft**
This PR fixes #14744 

using spread operator as method parameter is not safe, it might exceed the maximum stack size when dealing with a lot of records (130k in this case).

I have a PR for vscode and I have cherrypicked the changes in this PR.

### CODE REVIEW

File: src/sql/workbench/contrib/query/browser/messagePanel.ts [hunk 1]
```diff
@@ -201,7 +201,7 @@ export class MessagePanel extends Disposable {
 
 	private onMessage(message: IQueryMessage | IQueryMessage[], setInput: boolean = false) {
 		if (isArray(message)) {
-			this.model.messages.push(...message);
+			this.model.messages = this.model.messages.concat(message);
```
Comment (kburtram) [inline src/sql/workbench/contrib/query/browser/messagePanel.ts:204]: does this allocate and throw away an array for each message?  if so, is there any perf impact from the memory churn?

**Review by kburtram — APPROVED** [review]
LGTM.  though I'm not sure if this change is managing memory particularly efficiently if I'm reading the code correctly.

**Review by kisantia — APPROVED** [review]

**Comment by alanrenmsft** [comment]
I thought about it again and the problem is that we need an alternative way to do in place array push and splice that is safe for large arrays, so I implemented those and added it to the vscode arrays.ts. 

@kburtram Array.concat will return a new array which might not work for some scenarios, it might use more memory than the in place push but not a huge concern as we are not holding them and will be GC'ed by the runtime. 

while implementing the push/insert methods, I picked the approach of setting array length and then updating the values since it is faster comparing to other in place alternatives, it becomes more obvious as the number of items grow. here are the perf numbers, the test was done on https://jsbench.me/ 

2 array with 1000 items
image
2 array with 10,000 items
image
2 array with 100,000 items
image
2 array with 1,000,000 items
image

File: src/sql/workbench/contrib/query/browser/messagePanel.ts (hunk 1 above)
Comment (alanrenmsft) [reply src/sql/workbench/contrib/query/browser/messagePanel.ts]: not using concat in the new implementation anymore.

File: src/vs/base/common/arrays.ts [hunk 2]
```diff
@@ -612,3 +612,78 @@ export function mapFind<T, R>(array: Iterable<T>, mapFn: (value: T) => R | undef
 
 	return undefined;
 }
+
+/**
+ * An alternative for Array.push(...items) method, Use this if you need to push large number of items to the array.
```
Comment (Charles-Gagnon) [inline src/vs/base/common/arrays.ts:617]: Would suggest just putting these in our own arrays.ts class unless you're planning on trying to get VS Code to take it

**Review by Charles-Gagnon — APPROVED** [review]

File: src/vs/base/common/arrays.ts (hunk 2 above)
Comment (alanrenmsft) [reply src/vs/base/common/arrays.ts:617]: I need to use the splice method in a vscode file, here is the vscode PR https://github.com/microsoft/vscode/pull/119929

**Pull request merged**